File: not-proven/proven-tftp/ffi/zig/src/tftp.zig
// SPDX-License-Identifier: MPL-2.0
// Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <j.d.a.jewell@open.ac.uk>
//
// tftp.zig -- Zig FFI implementation of proven-tftp.
//
// Implements the TFTP (RFC 1350) transfer state machine with:
//   - 64-slot mutex-protected transfer session pool
//   - Read/write transfer state tracking
//   - Block number and retry count management
//   - Error code propagation
//   - Transfer mode validation
//   - Thread-safe via per-pool mutex
//
// All exported functions use C calling convention (callconv(.c)) and
// communicate state via u8 tags matching TFTPABI.Types exactly.

const std = @import("std");

// =========================================================================
// Enums (matching TFTPABI.Types tag assignments)
// =========================================================================

/// TFTP opcodes (ABI tags 0-4).
pub const Opcode = enum(u8) {
    rrq = 0,
    wrq = 1,
    data = 2,
    ack = 3,
    err = 4,
};

/// TFTP transfer modes (ABI tags 0-2).
pub const TransferMode = enum(u8) {
    netascii = 0,
    octet = 1,
    mail = 2,
};

/// TFTP error codes (ABI tags 0-7).
pub const TFTPError = enum(u8) {
    not_defined = 0,
    file_not_found = 1,
    access_violation = 2,
    disk_full = 3,
    illegal_operation = 4,
    unknown_tid = 5,
    file_exists = 6,
    no_such_user = 7,
};

/// Transfer states (ABI tags 0-4).
pub const TransferState = enum(u8) {
    idle = 0,
    reading = 1,
    writing = 2,
    in_error = 3,
    complete = 4,
};

// =========================================================================
// Internal data structures
// =========================================================================

/// Maximum concurrent sessions.
const MAX_SESSIONS: usize = 64;

/// Maximum filename length.
const MAX_FILENAME_LEN: usize = 256;

/// Maximum retries before giving up.
const MAX_RETRIES: u32 = 5;

/// TFTP block size (RFC 1350).
const BLOCK_SIZE: u32 = 512;

/// A TFTP transfer session.
const Session = struct {
    /// Current transfer state.
    state: TransferState,
    /// Filename being transferred.
    filename: [MAX_FILENAME_LEN]u8,
    filename_len: u32,
    /// Transfer mode.
    mode: TransferMode,
    /// Whether this is a read (true) or write (false) transfer.
    is_read: bool,
    /// Current block number.
    current_block: u16,
    /// Number of retries for the current block.
    retry_count: u32,
    /// Total bytes transferred.
    bytes_transferred: u32,
    /// Last error code (255 = no error).
    last_error: u8,
    /// Whether this session slot is in use.
    active: bool,
};

/// Default (empty) session.
const empty_session: Session = .{
    .state = .idle,
    .filename = [_]u8{0} ** MAX_FILENAME_LEN,
    .filename_len = 0,
    .mode = .octet,
    .is_read = true,
    .current_block = 0,
    .retry_count = 0,
    .bytes_transferred = 0,
    .last_error = 255,
    .active = false,
};

// =========================================================================
// Global state
// =========================================================================

var sessions: [MAX_SESSIONS]Session = [_]Session{empty_session} ** MAX_SESSIONS;
var mutex: std.Thread.Mutex = .{};

// =========================================================================
// Internal helpers
// =========================================================================

/// Validate a slot index, returning null if out of range or inactive.
fn validSlot(slot: c_int) ?usize {
    if (slot < 0 or slot >= MAX_SESSIONS) return null;
    const idx: usize = @intCast(slot);
    if (!sessions[idx].active) return null;
    return idx;
}

// =========================================================================
// Exported C ABI functions
// =========================================================================

/// Returns the ABI version number.
pub export fn tftp_abi_version() callconv(.c) u32 {
    return 1;
}

/// Create a new TFTP transfer session. Returns slot index (>=0) or -1 on failure.
/// is_read: 1 = read transfer, 0 = write transfer.
pub export fn tftp_create(
    filename_ptr: [*]const u8,
    filename_len: u32,
    mode: u8,
    is_read: u8,
) callconv(.c) c_int {
    mutex.lock();
    defer mutex.unlock();

    if (filename_len == 0 or filename_len > MAX_FILENAME_LEN) return -1;
    if (mode > 2) return -1;

    for (&sessions, 0..) |*s, i| {
        if (!s.active) {
            s.* = empty_session;
            @memcpy(s.filename[0..filename_len], filename_ptr[0..filename_len]);
            s.filename_len = filename_len;
            s.mode = @enumFromInt(mode);
            s.is_read = (is_read != 0);
            s.current_block = if (is_read != 0) 1 else 0;
            s.state = if (is_read != 0) .reading else .writing;
            s.active = true;
            return @intCast(i);
        }
    }
    return -1;
}

/// Destroy a session, releasing its slot.
pub export fn tftp_destroy(slot: c_int) callconv(.c) void {
    mutex.lock();
    defer mutex.unlock();
    if (slot < 0 or slot >= MAX_SESSIONS) return;
    sessions[@intCast(slot)] = empty_session;
}

/// Returns the current TransferState tag for a session.
pub export fn tftp_state(slot: c_int) callconv(.c) u8 {
    mutex.lock();
    defer mutex.unlock();
    const idx = validSlot(slot) orelse return 0;
    return @intFromEnum(sessions[idx].state);
}

/// Record receipt of a DATA block. Returns 0 on success, 1 on rejection.
pub export fn tftp_recv_data(
    slot: c_int,
    block_num: u16,
    data_len: u32,
    is_last: u8,
) callconv(.c) u8 {
    mutex.lock();
    defer mutex.unlock();

    const idx = validSlot(slot) orelse return 1;
    if (sessions[idx].state != .reading and sessions[idx].state != .writing) return 1;

    sessions[idx].current_block = block_num;
    sessions[idx].bytes_transferred += data_len;
    sessions[idx].retry_count = 0;

    if (is_last != 0) {
        sessions[idx].state = .complete;
    }
    return 0;
}

/// Record receipt of an ACK. Returns 0 on success, 1 on rejection.
pub export fn tftp_recv_ack(slot: c_int, block_num: u16) callconv(.c) u8 {
    mutex.lock();
    defer mutex.unlock();

    const idx = validSlot(slot) orelse return 1;
    if (sessions[idx].state != .writing and sessions[idx].state != .reading) return 1;

    sessions[idx].current_block = block_num;
    sessions[idx].retry_count = 0;
    return 0;
}

/// Record receipt of an ERROR. Returns 0 on success, 1 on rejection.
pub export fn tftp_recv_error(slot: c_int, error_code: u8) callconv(.c) u8 {
    mutex.lock();
    defer mutex.unlock();

    const idx = validSlot(slot) orelse return 1;
    if (sessions[idx].state == .idle or sessions[idx].state == .complete) return 1;
    if (error_code > 7) return 1;

    sessions[idx].last_error = error_code;
    sessions[idx].state = .in_error;
    return 0;
}

/// Increment retry counter. Returns 0 if retry ok, 1 if exhausted, 2 if rejected.
pub export fn tftp_retry(slot: c_int) callconv(.c) u8 {
    mutex.lock();
    defer mutex.unlock();

    const idx = validSlot(slot) orelse return 2;
    if (sessions[idx].state != .reading and sessions[idx].state != .writing) return 2;

    sessions[idx].retry_count += 1;
    if (sessions[idx].retry_count > MAX_RETRIES) {
        sessions[idx].state = .in_error;
        sessions[idx].last_error = 0; // Not defined
        return 1;
    }
    return 0;
}

/// Returns current block number.
pub export fn tftp_current_block(slot: c_int) callconv(.c) u16 {
    mutex.lock();
    defer mutex.unlock();
    const idx = validSlot(slot) orelse return 0;
    return sessions[idx].current_block;
}

/// Returns total bytes transferred.
pub export fn tftp_bytes_transferred(slot: c_int) callconv(.c) u32 {
    mutex.lock();
    defer mutex.unlock();
    const idx = validSlot(slot) orelse return 0;
    return sessions[idx].bytes_transferred;
}

/// Returns last error code (255 = no error).
pub export fn tftp_last_error(slot: c_int) callconv(.c) u8 {
    mutex.lock();
    defer mutex.unlock();
    const idx = validSlot(slot) orelse return 255;
    return sessions[idx].last_error;
}

/// Returns transfer mode tag.
pub export fn tftp_mode(slot: c_int) callconv(.c) u8 {
    mutex.lock();
    defer mutex.unlock();
    const idx = validSlot(slot) orelse return 0;
    return @intFromEnum(sessions[idx].mode);
}

/// Check if a transfer state transition is valid.
pub export fn tftp_can_transition(from: u8, to: u8) callconv(.c) u8 {
    if (from == 0 and to == 1) return 1; // Idle -> Reading
    if (from == 0 and to == 2) return 1; // Idle -> Writing
    if (from == 1 and to == 4) return 1; // Reading -> Complete
    if (from == 2 and to == 4) return 1; // Writing -> Complete
    if (from == 1 and to == 3) return 1; // Reading -> InError
    if (from == 2 and to == 3) return 1; // Writing -> InError
    return 0;
}

/// Check if a state is terminal (Complete or InError).
pub export fn tftp_is_terminal(state: u8) callconv(.c) u8 {
    if (state == 3 or state == 4) return 1; // InError or Complete
    return 0;
}

/// Returns number of active sessions.
pub export fn tftp_session_count() callconv(.c) u32 {
    mutex.lock();
    defer mutex.unlock();
    var count: u32 = 0;
    for (&sessions) |*s| {
        if (s.active) count += 1;
    }
    return count;
}

File: bindings/elixir/lib/proven_servers/tftp.ex
# SPDX-License-Identifier: MPL-2.0
# Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <j.d.a.jewell@open.ac.uk>

defmodule ProvenServers.Tftp do
  @moduledoc """
  TFTP (Trivial File Transfer Protocol) types for the proven-servers ABI.
  
  Mirrors the Idris2 module `TFTPABI.Types` and its type definitions:
  - `Opcode`        — TFTP opcodes (5 constructors, tags 0-4)
  - `TransferMode`  — TFTP transfer modes (3 constructors, tags 0-2)
  - `TftpError`     — TFTP error codes (8 constructors, tags 0-7)
  - `TransferState` — TFTP transfer lifecycle (5 constructors, tags 0-4)
  
  All discriminant values match the Idris2 ABI tag definitions exactly.

  All tag values match the Idris2 ABI definitions exactly.
  """

  @doc "Standard TFTP port (RFC 1350)."
  @spec tftp_port() :: non_neg_integer()
  def tftp_port, do: 69

  @doc "TFTP data block size (RFC 1350)."
  @spec tftp_block_size() :: non_neg_integer()
  def tftp_block_size, do: 512

  # ===========================================================================
  # Opcode (tags 0-4)
  # ===========================================================================

  @typedoc """
  Opcode types.

  Tag values match the Idris2 ABI definitions exactly.
  """
  @type opcode :: :rrq | :wrq | :data | :ack | :error

  @opcode_tags %{
    rrq: 0,
    wrq: 1,
    data: 2,
    ack: 3,
    error: 4,
  }

  @tag_to_opcode Map.new(@opcode_tags, fn {k, v} -> {v, k} end)

  @doc """
  Decode a `Opcode` from the C-ABI tag value.

  Returns `{:ok, atom}` for valid tags 0..4, `:error` for invalid.

  ## Examples

      iex> ProvenServers.Tftp.opcode_from_tag(0)
      {:ok, :rrq}
  """
  @spec opcode_from_tag(non_neg_integer()) :: {:ok, opcode()} | :error
  def opcode_from_tag(tag) when is_integer(tag) and tag >= 0 and tag <= 4 do
    {:ok, Map.fetch!(@tag_to_opcode, tag)}
  end

  def opcode_from_tag(_tag), do: :error

  @doc """
  Encode a `Opcode` to the C-ABI tag value.
  """
  @spec opcode_to_tag(opcode()) :: non_neg_integer()
  def opcode_to_tag(val) when is_map_key(@opcode_tags, val) do
    Map.fetch!(@opcode_tags, val)
  end

  @doc """
  All `Opcode` variants in tag order.
  """
  @spec all_opcodes() :: [opcode()]
  def all_opcodes, do: [:rrq, :wrq, :data, :ack, :error]

  @doc """
  Decode from an ABI tag value.

  Encode to the ABI tag value.

  Whether this opcode initiates a transfer.
  """
  @spec is_request?(opcode()) :: boolean()
  def is_request?(val) when val in [:rrq, :wrq], do: true
  def is_request?(_val), do: false

  @doc """
  Whether this opcode carries payload data.
  """
  @spec is_data?(opcode()) :: boolean()
  def is_data?(val) when val in [:data], do: true
  def is_data?(_val), do: false

  # ===========================================================================
  # TransferMode (tags 0-2)
  # ===========================================================================

  @typedoc """
  TransferMode types.

  Tag values match the Idris2 ABI definitions exactly.
  """
  @type transfer_mode :: :net_ascii | :octet | :mail

  @transfer_mode_tags %{
    net_ascii: 0,
    octet: 1,
    mail: 2,
  }

  @tag_to_transfer_mode Map.new(@transfer_mode_tags, fn {k, v} -> {v, k} end)

  @doc """
  Decode a `TransferMode` from the C-ABI tag value.

  Returns `{:ok, atom}` for valid tags 0..2, `:error` for invalid.

  ## Examples

      iex> ProvenServers.Tftp.transfer_mode_from_tag(0)
      {:ok, :net_ascii}
  """
  @spec transfer_mode_from_tag(non_neg_integer()) :: {:ok, transfer_mode()} | :error
  def transfer_mode_from_tag(tag) when is_integer(tag) and tag >= 0 and tag <= 2 do
    {:ok, Map.fetch!(@tag_to_transfer_mode, tag)}
  end

  def transfer_mode_from_tag(_tag), do: :error

  @doc """
  Encode a `TransferMode` to the C-ABI tag value.
  """
  @spec transfer_mode_to_tag(transfer_mode()) :: non_neg_integer()
  def transfer_mode_to_tag(val) when is_map_key(@transfer_mode_tags, val) do
    Map.fetch!(@transfer_mode_tags, val)
  end

  @doc """
  All `TransferMode` variants in tag order.
  """
  @spec all_transfer_modes() :: [transfer_mode()]
  def all_transfer_modes, do: [:net_ascii, :octet, :mail]

  @doc """
  Decode from an ABI tag value.

  Encode to the ABI tag value.

  The TFTP mode string (case-insensitive per RFC).
        match self {

  Whether this mode performs character set conversion.
  """
  @spec is_text_mode?(transfer_mode()) :: boolean()
  def is_text_mode?(val) when val in [:net_ascii], do: true
  def is_text_mode?(_val), do: false

  @doc """
  Whether this transfer mode is deprecated.
  """
  @spec is_deprecated?(transfer_mode()) :: boolean()
  def is_deprecated?(val) when val in [:mail], do: true
  def is_deprecated?(_val), do: false

  # ===========================================================================
  # TftpError (tags 0-7)
  # ===========================================================================

  @typedoc """
  TftpError types.

  Tag values match the Idris2 ABI definitions exactly.
  """
  @type tftp_error ::
          :not_defined
          | :file_not_found
          | :access_violation
          | :disk_full
          | :illegal_operation
          | :unknown_tid
          | :file_exists
          | :no_such_user

  @tftp_error_tags %{
    not_defined: 0,
    file_not_found: 1,
    access_violation: 2,
    disk_full: 3,
    illegal_operation: 4,
    unknown_tid: 5,
    file_exists: 6,
    no_such_user: 7,
  }

  @tag_to_tftp_error Map.new(@tftp_error_tags, fn {k, v} -> {v, k} end)

  @doc """
  Decode a `TftpError` from the C-ABI tag value.

  Returns `{:ok, atom}` for valid tags 0..7, `:error` for invalid.

  ## Examples

      iex> ProvenServers.Tftp.tftp_error_from_tag(0)
      {:ok, :not_defined}
  """
  @spec tftp_error_from_tag(non_neg_integer()) :: {:ok, tftp_error()} | :error
  def tftp_error_from_tag(tag) when is_integer(tag) and tag >= 0 and tag <= 7 do
    {:ok, Map.fetch!(@tag_to_tftp_error, tag)}
  end

  def tftp_error_from_tag(_tag), do: :error

  @doc """
  Encode a `TftpError` to the C-ABI tag value.
  """
  @spec tftp_error_to_tag(tftp_error()) :: non_neg_integer()
  def tftp_error_to_tag(val) when is_map_key(@tftp_error_tags, val) do
    Map.fetch!(@tftp_error_tags, val)
  end

  @doc """
  All `TftpError` variants in tag order.
  """
  @spec all_tftp_errors() :: [tftp_error()]
  def all_tftp_errors do
    [
      :not_defined, :file_not_found, :access_violation, :disk_full, :illegal_operation,
      :unknown_tid, :file_exists, :no_such_user
    ]
  end

  @doc """
  Decode from an ABI tag value.

  Encode to the ABI tag value.

  Whether this error relates to access control.
  """
  @spec is_access_error?(tftp_error()) :: boolean()
  def is_access_error?(val) when val in [:access_violation, :no_such_user], do: true
  def is_access_error?(_val), do: false

  @doc """
  Whether this error relates to storage capacity.
  """
  @spec is_storage_error?(tftp_error()) :: boolean()
  def is_storage_error?(val) when val in [:disk_full, :file_exists], do: true
  def is_storage_error?(_val), do: false

  # ===========================================================================
  # TransferState (tags 0-4)
  # ===========================================================================

  @typedoc """
  TransferState types.

  Tag values match the Idris2 ABI definitions exactly.
  """
  @type transfer_state :: :idle | :reading | :writing | :in_error | :complete

  @transfer_state_tags %{
    idle: 0,
    reading: 1,
    writing: 2,
    in_error: 3,
    complete: 4,
  }

  @tag_to_transfer_state Map.new(@transfer_state_tags, fn {k, v} -> {v, k} end)

  @doc """
  Decode a `TransferState` from the C-ABI tag value.

  Returns `{:ok, atom}` for valid tags 0..4, `:error` for invalid.

  ## Examples

      iex> ProvenServers.Tftp.transfer_state_from_tag(0)
      {:ok, :idle}
  """
  @spec transfer_state_from_tag(non_neg_integer()) :: {:ok, transfer_state()} | :error
  def transfer_state_from_tag(tag) when is_integer(tag) and tag >= 0 and tag <= 4 do
    {:ok, Map.fetch!(@tag_to_transfer_state, tag)}
  end

  def transfer_state_from_tag(_tag), do: :error

  @doc """
  Encode a `TransferState` to the C-ABI tag value.
  """
  @spec transfer_state_to_tag(transfer_state()) :: non_neg_integer()
  def transfer_state_to_tag(val) when is_map_key(@transfer_state_tags, val) do
    Map.fetch!(@transfer_state_tags, val)
  end

  @doc """
  All `TransferState` variants in tag order.
  """
  @spec all_transfer_states() :: [transfer_state()]
  def all_transfer_states, do: [:idle, :reading, :writing, :in_error, :complete]

  @doc """
  Decode from an ABI tag value.

  Encode to the ABI tag value.

  Whether a transfer is actively in progress.
  """
  @spec is_active?(transfer_state()) :: boolean()
  def is_active?(val) when val in [:reading, :writing], do: true
  def is_active?(_val), do: false

  @doc """
  Whether the transfer has reached a terminal state.
  """
  @spec is_terminal?(transfer_state()) :: boolean()
  def is_terminal?(val) when val in [:in_error, :complete], do: true
  def is_terminal?(_val), do: false

end
